### PR TITLE
[AIEX] Add symbol-based IpuDmaWaitOp

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a new AIEX `IpuDmaWaitOp`, which uses a global symbol to retrieve the necessary physical synchronization info like `column`, `channel`, `direction`. This new op is converted into `IpuSyncOp` inside the `DmaToIpu` pass. `IpuDmaWaitOp` should be easier to use than `IpuSyncOp` and abstracts away physical details as the latter requires users to already be aware of physical information, which might actually not be available yet. For example, a lot of reference designs use `IpuSyncOp` with channel set to `0` for output synchronization, however, in principal channel `1` could also be used and in general the exact channel has not been determined at the stage these reference designs intercept. 

See for example the syncs here: https://github.com/Xilinx/mlir-aie/blob/a74813ababb3d9631dc5483a94a10e6f2ffe401d/programming_examples/basic/matrix_multiplication/matrix_vector/aie2.py#L207, which use channel `0`, however, the operations to be synchronized don't reference a channel `0` yet and the exact channel hasn't been determined yet.

### CODE REVIEW

File: include/aie/Dialect/AIEX/IR/AIEX.td [hunk 1]
```diff
@@ -518,6 +519,36 @@ def AIE_IpuDmaMemcpyNdOp: AIEX_Op<"ipu.dma_memcpy_nd", [
   let hasVerifier = 1;
 }
 
+def AIE_IpuDmaWaitOp: AIEX_Op<"ipu.dma_wait", []> {
+  let summary = "Blocking operation to wait for a DMA to complete execution.";
+  let description = [{
+    The IpuDmaWaitOp blocks until the DMA referenced through `symbol` completes execution
+    and issues a task-complete-token. NOTE(jornt): do we want to support synchronization
```
Comment (makslevental) [inline include/aie/Dialect/AIEX/IR/AIEX.td:526]: drive-by comments (will review thoroughly soon): I believe this note should be removed?

File: lib/Dialect/AIEX/IR/AIEXDialect.cpp [hunk 2]
```diff
@@ -105,6 +105,16 @@ LogicalResult AIEX::IpuDmaMemcpyNdOp::verify() {
   return success();
 }
 
+LogicalResult AIEX::IpuDmaWaitOp::verify() {
+  auto dev = (*this)->getParentOfType<AIE::DeviceOp>();
+  if (!dev)
+    return emitOpError("couldn't find parent of type DeviceOp");
+  auto sym = dev.lookupSymbol(getSymbol());
```
Comment (makslevental) [inline lib/Dialect/AIEX/IR/AIEXDialect.cpp:112]: same

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp [hunk 3]
```diff
@@ -335,6 +335,37 @@ struct DmaToIpuPattern : OpConversionPattern<IpuDmaMemcpyNdOp> {
...
+    if (!shimDmaAllocOp) {
+      op.emitOpError("couldn't find shim_dma_allocation op");
+      return failure();
+    }
+    auto channelDir = shimDmaAllocOp->getChannelDir();
+    int channel = shimDmaAllocOp->getChannelIndex();
+    int direction = (int)(channelDir == AIE::DMAChannelDir::MM2S);
+    int column = shimDmaAllocOp->getCol();
+
+    // Create with `column_num == 1` and `row_num == 1` to check for a single
```
Comment (makslevental) [inline lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:361]: I guess this is reasonable (always setting both of these to 1) but I think the signal here is that actually `IpuSyncOp` just shouldn't expose these attributes (and then in the write32s set these to 1).

**Review by makslevental — COMMENTED** [review]
Generally looks good - will come back (soon) for a closer look after it has settled in my brain...

File: include/aie/Dialect/AIEX/IR/AIEX.td (hunk 1 above)
Comment (jtuyls) [reply include/aie/Dialect/AIEX/IR/AIEX.td:526]: Yeah, I just wanted to bring this up. Is there any polling support right now or any interest in adding that?

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 3 above)
Comment (jtuyls) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:361]: Maybe it's ok in `IpuSyncOp`. I have to look at the details again, but you could use this as a `wait_all` I think? For example increase `col_num` to wait for multiple TCTs from different columns? A `DmaWaitAllOp` could be added to accommodate that if needed.

File: include/aie/Dialect/AIEX/IR/AIEX.td (hunk 1 above)
Comment (makslevental) [reply include/aie/Dialect/AIEX/IR/AIEX.td]: I don't know 100% but just yesterday I was discussing this with @benvanik - there is indeed interest and I will likely figure the how/when/where/why soon.

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 3 above)
Comment (jtuyls) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:361]: Ideally, `ipu_sync` wouldn't be used anymore directly as `dma_wait` should be more accessible and then `column_num` and `row_num` would only be set in a couple of places. We could also make these arguments optional and default to 1.

Comment (fifield) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:361]: There's a similar `column_num` field in the `writebd_shimtile` instruction. None of these map to bits in hardware registers. Instead they are a firmware feature to write to multiple tiles with one instruction. I don't know of anyone using the them with mlir-aie.

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp [hunk 4]
```diff
@@ -335,6 +335,37 @@ struct DmaToIpuPattern : OpConversionPattern<IpuDmaMemcpyNdOp> {
...
+struct DmaWaitToIpuPattern : OpConversionPattern<IpuDmaWaitOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  DmaWaitToIpuPattern(MLIRContext *context, PatternBenefit benefit = 1)
+      : OpConversionPattern(context, benefit) {}
+
+  LogicalResult
+  matchAndRewrite(IpuDmaWaitOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto dev = op->getParentOfType<AIE::DeviceOp>();
```
Comment (makslevental) [inline lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:350]: auto

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp [hunk 5]
```diff
@@ -335,6 +335,37 @@ struct DmaToIpuPattern : OpConversionPattern<IpuDmaMemcpyNdOp> {
...
+    auto shimDmaAllocOp = getAllocOpForSymbol(dev, op.getSymbol());
```
Comment (makslevental) [inline lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:351]: auto

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp [hunk 6]
```diff
@@ -335,6 +335,37 @@ struct DmaToIpuPattern : OpConversionPattern<IpuDmaMemcpyNdOp> {
...
+    if (!shimDmaAllocOp) {
+      op.emitOpError("couldn't find shim_dma_allocation op");
+      return failure();
+    }
+    auto channelDir = shimDmaAllocOp->getChannelDir();
```
Comment (makslevental) [inline lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp:356]: auto

**Review by makslevental — APPROVED** [review]
LGTM modulo the few remaining `auto`s - please fix and then feel free to merge.

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 4 above)
Comment (jtuyls) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp]: Changed it, but I think this would be an acceptable case in upstream as well as the type is clear from `<AIE::DeviceOp>`

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 5 above)
Comment (jtuyls) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp]: Done

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 6 above)
Comment (jtuyls) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp]: Done

File: lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp (hunk 4 above)
Comment (makslevental) [reply lib/Dialect/AIEX/Transforms/AIEDmaToIpu.cpp]: Eh one reason I prefer the full type is because with these templated members sometimes you actually get a pointer rather a value so it doesn't actually match what's inside `<...>`.
